### PR TITLE
Add activity after death

### PR DESCRIPTION
**Pull request by rchlv**
Related to #862
WIP - Added function for get_all_slf_deaths_lookup_path which combines all financial years' extract.
       - Added function for add_activity_after_death_flag

Following identification of activity recorded after death due to data quality or admin issues, an investigation into sources of death data available for SLFs was undertaken. This guided the addition of a new variable to flag records where activity occurred after the death of date. Further details about the issues and methodology can be found in the accompanying "Investigation into activity after death" report.

### CODE REVIEW

**Review by SwiftySalmon — COMMENTED** [review]
code looks fine! Need to run it tomorrow and see how the output looks

File: R/add_activity_after_death_flag.R [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+#'
+#' @param data episode files
+#' @param year financial year, e.g. '1920'
+#' @param deaths_data The death data for the year
+#'
+#' @return data flagged if activity after death
+add_activity_after_death_flag <- function(
+    data,
+    year,
+    deaths_data = read_file(all_slf_deaths_lookup_path())) {
```
Comment (SwiftySalmon) [inline R/add_activity_after_death_flag.R:14]: Is this the correct function name?

File: R/add_activity_after_death_flag.R [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    dplyr::filter(record_keydate1 > death_date_boxi)
+
+  if (nrow(check_death_date_match) != 0) {
+    warning("There were records in the episode file which have a record_keydate1 after the BOXI NRS death date.")
+  }
+
+
+  flag_data <- data %>%
+    dplyr::mutate(
+      flag_keydate1 = if_else(record_keydate1 > death_date_boxi, 1, 0),
```
Comment (SwiftySalmon) [inline R/add_activity_after_death_flag.R:45]: if_else statements need dplyr::

File: R/add_activity_after_death_flag.R [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+      # Also flag records without a death_date in the episode file, but the BOXI death date occurs in the current or a previous financial year.
+      flag_deathdate_missing = if_else(((is.na(death_date) | death_date == "") & (death_date_boxi <= paste0("20", substr(year, 3, 4), "-03-31"))), 1, 0)
+    ) %>%
+    # These should be flagged by one of the two lines of code above, but in these cases, we will also fill in the blank death date if appropriate
+
+    # Search all variables beginning with "flag_" for value "1" and create new variable to flag cases where 1 is present
+    # Multiplying by 1 changes flag from true/false to 1/0
+    dplyr::mutate(activity_after_death = purrr::pmap_dbl(
+      select(., contains("flag_")),
```
Comment (SwiftySalmon) [inline R/add_activity_after_death_flag.R:60]: dplyr::select

File: R/add_activity_after_death_flag.R [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+  if (nrow(check_deceased_match) != 0) {
+    warning("There were records in the episode file which have a deceased variable which does not match the BOXI NRS deceased variable")
+  }
+
+
+  # Fill in date of death if missing in the episode file but available in BOXI lookup, due to historic dates of death not being carried
+  # over from previous financial years
+  flag_data <- flag_data %>%
+    dplyr::mutate(death_date = if_else(((is.na(death_date) | death_date == "") & (death_date_boxi <= paste0("20", substr(year, 1, 2), "-03-31"))), death_date_boxi, death_date)) %>%
```
Comment (SwiftySalmon) [inline R/add_activity_after_death_flag.R:80]: dplyr::

File: R/add_activity_after_death_flag.R [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+#' @param year financial year, e.g. '1920'
+#' @param deaths_data The death data for the year
+#'
+#' @return data flagged if activity after death
+add_activity_after_death_flag <- function(
+    data,
+    year,
+    deaths_data = read_file(all_slf_deaths_lookup_path())) {
+  # Match on BOXI NRS deaths lookup for records without chi
+  data <- data %>%
```
Comment (SwiftySalmon) [inline R/add_activity_after_death_flag.R:16]: I think this should be named something else. the join at the end isn't working for me

Comment (SwiftySalmon) [reply R/add_activity_after_death_flag.R:16]: maybe something like 
death_activity_lookup <- data %>%
    dplyr::select(year, anon_chi, record_keydate1, record_keydate2,death_date, deceased) %>%....

File: R/add_activity_after_death_flag.R [hunk 6]
```diff
@@ -11,19 +11,22 @@
 add_activity_after_death_flag <- function(
     data,
     year,
-    deaths_data = read_file(all_slf_deaths_lookup_path())) {
-  # Match on BOXI NRS deaths lookup for records without chi
-  data <- data %>%
+    deaths_data = read_file(get_all_slf_deaths_lookup_path())) {
```
Comment (Jennit07) [inline R/add_activity_after_death_flag.R:14]: i think the `slfhelper::get_chi()` should go here 

deaths_data = read_file(get_all_slf_deaths_lookup_path()) %>% slfhelper::get_chi(())

File: R/add_activity_after_death_flag.R [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+      # Also flag records without a death_date in the episode file, but the BOXI death date occurs in the current or a previous financial year.
+      flag_deathdate_missing = dplyr::if_else(((is.na(death_date) | death_date == "") & (death_date_boxi <= paste0("20", substr(year, 3, 4), "-03-31"))), 1, 0)
+    ) %>%
+    # These should be flagged by one of the two lines of code above, but in these cases, we will also fill in the blank death date if appropriate
+
+    # Search all variables beginning with "flag_" for value "1" and create new variable to flag cases where 1 is present
+    # Multiplying by 1 changes flag from true/false to 1/0
+    dplyr::mutate(activity_after_death = purrr::pmap_dbl(
+      dplyr::select(., contains("flag_")),
```
Comment (Jennit07) [inline R/add_activity_after_death_flag.R:63]:
```suggestion
      dplyr::select(., tidyselect::contains("flag_")),
```

File: R/add_activity_after_death_flag.R [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+#' @param year financial year, e.g. '1920'
+#' @param deaths_data The death data for the year
+#'
+#' @return data flagged if activity after death
+add_activity_after_death_flag <- function(
+    data,
+    year,
+    deaths_data = read_file(get_all_slf_deaths_lookup_path())) {
+  death_joined <- data %>%
+    dplyr::select(year, chi, record_keydate1, record_keydate2, death_date, deceased) %>%
```
Comment (Jennit07) [inline R/add_activity_after_death_flag.R:16]:
```suggestion
    dplyr::select(.data$year, .data$chi, .data$record_keydate1, .data$record_keydate2, .data$death_date, .data$deceased) %>%
```
I think we need `.data$` throughout where we are using variables

Comment (Jennit07) [reply R/add_activity_after_death_flag.R:16]: or if this has depreciated with one of the versions of tidyverse then we need quotes in place

**Pull request merged**